### PR TITLE
feat: [CDS-34991]: Pipeline Templates

### PR DESCRIPTION
**Pull request by yc147**
##### Screenshots:

You can use the following comments to re-trigger PR Checks

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier - `fix prettier`

### CODE REVIEW

File: src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineSpecifications/TemplatePipelineSpecifications.tsx [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+  React.useEffect(() => {
+    try {
+      const templateInputs = parse(defaultTo(templateInputSetYaml?.data, ''))
+      setInputsTemplate(templateInputs)
+      setFormikErrors({})
+    } catch (error) {
+      showError(error.message, undefined, 'template.parse.inputSet.error')
+    }
+  }, [templateInputSetYaml?.data])
```
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineSpecifications/TemplatePipelineSpecifications.tsx:106]: Codacy has a fix for the issue: [React Hook React.useEffect has a missing dependency: 'showError'. Either include it or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
  }, [showError, templateInputSetYaml?.data])
```

File: src/modules/70-pipeline/components/PipelineStudio/SavePipelinePopover/SavePipelinePopover.tsx [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+        return
+      }
+      setSelection({ stageId, stepId })
+      updatePipelineView({
+        ...pipelineView,
+        isSplitViewOpen: true,
+        splitViewData: { type: SplitViewTypes.StageView }
+      })
+    },
+    [isYaml, pipelineView]
```
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/SavePipelinePopover/SavePipelinePopover.tsx:165]: Codacy has a fix for the issue: [React Hook React.useCallback has missing dependencies: 'hideErrorModal', 'setSelection', and 'updatePipelineView'. Either include them or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
    [hideErrorModal, isYaml, pipelineView, setSelection, updatePipelineView]
```

File: src/modules/70-pipeline/components/PipelineStudio/SavePipelinePopover/SavePipelinePopover.tsx [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+  const { showSuccess, showError, clear } = useToaster()
+  const { getString } = useStrings()
+  const { OPA_PIPELINE_GOVERNANCE } = useFeatureFlags()
+  const history = useHistory()
+  const { projectIdentifier, orgIdentifier, accountId, pipelineIdentifier, module } =
+    useParams<PipelineType<PipelinePathProps>>()
+  const isYaml = view === SelectedView.YAML
+  const [menuOpen, setMenuOpen] = React.useState(false)
+  const templatesFeatureFlagEnabled = useFeatureFlag(FeatureFlag.NG_TEMPLATES)
+  const [updatePipelineAPIResponse, setUpdatePipelineAPIResponse] = React.useState<any>()
```
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/SavePipelinePopover/SavePipelinePopover.tsx:112]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
  const [updatePipelineAPIResponse, setUpdatePipelineAPIResponse] = React.useState<unknown>()
```

File: src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineSpecifications/TemplatePipelineSpecifications.tsx [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+      setInitialValues(cloneDeep(pipeline))
+    } catch (error) {
+      showError(error.message, undefined, 'template.parse.inputSet.error')
+    }
+  }, [inputsTemplate])
+
+  React.useEffect(() => {
+    dummyPipeline.current = pipeline
+    setFormikErrors({})
+  }, [pipeline.template?.templateRef, pipeline.template?.versionLabel])
```
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineSpecifications/TemplatePipelineSpecifications.tsx:129]: Codacy has a fix for the issue: [React Hook React.useEffect has a missing dependency: 'pipeline'. Either include it or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
  }, [pipeline, pipeline.template.templateRef, pipeline.template.versionLabel])
```

File: src/modules/70-pipeline/components/PipelineStudio/SavePipelinePopover/SavePipelinePopover.tsx (hunk 3 above)
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/SavePipelinePopover/SavePipelinePopover.tsx:112]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
  const [updatePipelineAPIResponse, setUpdatePipelineAPIResponse] = React.useState<unknown>()
```

File: src/modules/70-pipeline/components/PipelineStudio/PipelineVariables/PipelineVariables.tsx [hunk 5]
```diff
@@ -53,23 +54,111 @@ export function PipelineVariables(): React.ReactElement {
...
+  )
+
+  const updateStage = React.useCallback(
+    async (values: StageElementConfig) => {
+      if (pipeline.stages) {
+        set(pipeline, 'stages', updateStages(values, pipeline.stages))
+        updatePipeline(pipeline)
+      }
+    },
+    [pipeline, updatePipeline]
```
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineVariables/PipelineVariables.tsx:143]: Codacy has a fix for the issue: [React Hook React.useCallback has a missing dependency: 'updateStages'. Either include it or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
    [pipeline, updatePipeline, updateStages]
```

File: src/modules/70-pipeline/components/PipelineStudio/SavePipelinePopover/SavePipelinePopover.tsx [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+  const { getRBACErrorMessage } = useRBACError()
+  const { getString } = useStrings()
+  const { OPA_PIPELINE_GOVERNANCE } = useFeatureFlags()
+  const history = useHistory()
+  const { projectIdentifier, orgIdentifier, accountId, pipelineIdentifier, module } =
+    useParams<PipelineType<PipelinePathProps>>()
+  const isYaml = view === SelectedView.YAML
+  const [menuOpen, setMenuOpen] = React.useState(false)
+  const templatesFeatureFlagEnabled = useFeatureFlag(FeatureFlag.NG_TEMPLATES)
+  const [updatePipelineAPIResponse, setUpdatePipelineAPIResponse] = React.useState<any>()
```
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/SavePipelinePopover/SavePipelinePopover.tsx:115]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
  const [updatePipelineAPIResponse, setUpdatePipelineAPIResponse] = React.useState<unknown>()
```

File: src/modules/70-pipeline/components/PipelineStudio/PipelineVariables/PipelineVariables.tsx (hunk 5 above)
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineVariables/PipelineVariables.tsx:143]: Codacy has a fix for the issue: [React Hook React.useCallback has a missing dependency: 'updateStages'. Either include it or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
    [pipeline, updatePipeline, updateStages]
```

File: src/modules/72-templates-library/components/TemplateStudio/PipelineTemplateCanvas/PipelineTemplateCanvas.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    identifier: DefaultNewPipelineId
+  })
+
+  const [pipeline, setPipeline] = React.useState<PipelineInfoConfig>(createPipelineFromTemplate())
+
+  React.useEffect(() => {
+    if (!isLoading && !isUpdated) {
+      setPipeline(createPipelineFromTemplate())
+    }
+  }, [isLoading, isUpdated])
```
Comment (diptiman-adak) [inline src/modules/72-templates-library/components/TemplateStudio/PipelineTemplateCanvas/PipelineTemplateCanvas.tsx:54]: Codacy has a fix for the issue: [React Hook React.useEffect has a missing dependency: 'createPipelineFromTemplate'. Either include it or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
  }, [createPipelineFromTemplate, isLoading, isUpdated])
```

File: src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineSpecifications/TemplatePipelineSpecifications.tsx [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+  React.useEffect(() => {
+    try {
+      const mergedTemplateInputs = merge({}, inputsTemplate, pipeline.template?.templateInputs)
+      setTemplateInputs(pipeline, mergedTemplateInputs)
+      updatePipeline(pipeline)
+      setInitialValues(cloneDeep(pipeline))
+    } catch (error) {
+      showError(error.message, undefined, 'template.parse.inputSet.error')
+    }
+  }, [inputsTemplate])
```
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineSpecifications/TemplatePipelineSpecifications.tsx:124]: Codacy has a fix for the issue: [React Hook React.useEffect has missing dependencies: 'pipeline', 'showError', and 'updatePipeline'. Either include them or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
  }, [inputsTemplate, pipeline, showError, updatePipeline])
```

File: src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineSpecifications/TemplatePipelineSpecifications.tsx [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+  React.useEffect(() => {
+    if (!templateInputSetLoading) {
+      try {
+        const templateInputs = parse(defaultTo(templateInputSetYaml?.data, ''))
+        setInputsTemplate(templateInputs)
+      } catch (error) {
+        showError(error.message, undefined, 'template.parse.inputSet.error')
+      }
+    }
+  }, [templateInputSetLoading, templateInputSetYaml?.data])
```
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineSpecifications/TemplatePipelineSpecifications.tsx:107]: Codacy has a fix for the issue: [React Hook React.useEffect has a missing dependency: 'showError'. Either include it or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
  }, [showError, templateInputSetLoading, templateInputSetYaml?.data])
```

File: src/modules/70-pipeline/components/PipelineStudio/SavePipelinePopover/SavePipelinePopover.tsx [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+
+    // if Git sync enabled then display modal
+    if (isGitSyncEnabled) {
+      if (isEmpty(gitDetails.repoIdentifier) || isEmpty(gitDetails.branch)) {
+        clear()
+        showError(getString('pipeline.gitExperience.selectRepoBranch'))
+        return
+      }
+      // When git sync enabled, do not irritate user by taking all git info then at the end showing BE errors related to schema
+      const error = await validateJSONWithSchema({ pipeline: latestPipeline }, pipelineSchema?.data as any)
```
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/SavePipelinePopover/SavePipelinePopover.tsx:326]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
      const error = await validateJSONWithSchema({ pipeline: latestPipeline }, pipelineSchema?.data as unknown)
```

File: src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineCanvas/TemplatePipelineCanvas.tsx [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+  React.useEffect(() => {
+    if (pipelineResponse?.data?.mergedPipelineYaml) {
+      setResolvedPipeline(parse(pipelineResponse?.data?.mergedPipelineYaml))
+    }
+  }, [pipelineResponse?.data?.mergedPipelineYaml])
+
+  React.useEffect(() => {
+    dummyPipeline.current = pipeline
+  }, [pipeline.template?.templateRef, pipeline.template?.versionLabel])
```
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineCanvas/TemplatePipelineCanvas.tsx:94]: Codacy has a fix for the issue: [React Hook React.useEffect has a missing dependency: 'pipeline'. Either include it or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
  }, [pipeline, pipeline.template.templateRef, pipeline.template.versionLabel])
```

File: src/modules/70-pipeline/components/PipelineStudio/StepUtil.ts [hunk 12]
```diff
@@ -418,78 +418,97 @@ export const validatePipeline = ({
...
       }
     }
-  }
 
-  if (pipeline?.variables) {
-    const step = factory.getStep(StepType.CustomVariable)
-    const errorsResponse: any = step?.validateInputSet({ data: pipeline, template, getString, viewType })
+    if (pipeline?.variables) {
+      const step = factory.getStep(StepType.CustomVariable)
+      const errorsResponse: any = step?.validateInputSet({ data: pipeline, template, getString, viewType })
```
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/StepUtil.ts:467]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
    const errorsResponse: unknown = step?.validateInputSet({ data: pipeline, template, getString, viewType })
```

File: src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineSpecifications/TemplatePipelineSpecifications.tsx (hunk 8 above)
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineSpecifications/TemplatePipelineSpecifications.tsx:124]: Codacy has a fix for the issue: [React Hook React.useEffect has missing dependencies: 'pipeline', 'showError', and 'updatePipeline'. Either include them or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
  }, [inputsTemplate, pipeline, showError, updatePipeline])
```

File: src/modules/70-pipeline/components/PipelineStudio/PipelineVariables/PipelineVariables.tsx [hunk 13]
```diff
@@ -146,23 +127,117 @@ export function PipelineVariablesWithRef(
...
   }
 
+  return (
+    <div className={css.pipelineVariables}>
+      {error ? (
+        <PageError message={(error?.data as Error)?.message || error?.message} />
+      ) : (
+        <div className={css.content}>
+          <VariablesHeader applyChanges={applyChanges} discardChanges={discardChanges} />
+          <div className={css.variableList} ref={pipelineVariablesRef as any}>
```
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineVariables/PipelineVariables.tsx:137]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
          <div className={css.variableList} ref={pipelineVariablesRef as unknown}>
```

File: src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineSpecifications/TemplatePipelineSpecifications.tsx (hunk 9 above)
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineTemplateBuilder/TemplatePipelineSpecifications/TemplatePipelineSpecifications.tsx:107]: Codacy has a fix for the issue: [React Hook React.useEffect has a missing dependency: 'showError'. Either include it or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
  }, [showError, templateInputSetLoading, templateInputSetYaml?.data])
```

File: src/modules/70-pipeline/components/PipelineStudio/SavePipelinePopover/SavePipelinePopover.tsx [hunk 14]
```diff
@@ -0,0 +1,453 @@
...
+        return
+      }
+      setSelection({ stageId, stepId })
+      updatePipelineView({
+        ...pipelineView,
+        isSplitViewOpen: true,
+        splitViewData: { type: SplitViewTypes.StageView }
+      })
+    },
+    [isYaml, pipelineView]
```
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/SavePipelinePopover/SavePipelinePopover.tsx:168]: Codacy has a fix for the issue: [React Hook React.useCallback has missing dependencies: 'hideErrorModal', 'setSelection', and 'updatePipelineView'. Either include them or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
    [hideErrorModal, isYaml, pipelineView, setSelection, updatePipelineView]
```

File: src/modules/70-pipeline/components/PipelineStudio/PipelineVariables/PipelineVariables.tsx [hunk 15]
```diff
@@ -146,23 +127,117 @@ export function PipelineVariablesWithRef(
...
+  )
+
+  const updateStage = React.useCallback(
+    async (values: StageElementConfig) => {
+      if (pipeline.stages) {
+        set(pipeline, 'stages', updateStages(values, pipeline.stages))
+        updatePipeline(pipeline)
+      }
+    },
+    [pipeline, updatePipeline]
```
Comment (diptiman-adak) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineVariables/PipelineVariables.tsx:221]: Codacy has a fix for the issue: [React Hook React.useCallback has a missing dependency: 'updateStages'. Either include it or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232314)
```suggestion
    [pipeline, updatePipeline, updateStages]
```

**Review by kapilkumawat86 — APPROVED** [review]

**Pull request merged**